### PR TITLE
Multithreading Render Passes Sample

### PR DESCRIPTION
**Pull request by PavelRudko-ARM**
## Description

This sample shows how we can benifit by recording different render passes in multiple threads.

Please see [tutorial](https://github.com/KhronosGroup/Vulkan-Samples/blob/4f3a7b3c8b7f31dfd1479d66fca908d7f802aff4/samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md).

This PR adds the following framework changes:

-  Orthographic camera (in this case to render the shadow map)
- `CommandBuffer` now allows to specify render pass and framebuffer in `begin` and `begin_renderpass`
- `RenderContext` allows to submit multiple command buffers at once
- 'GeometrySubpass' provides more control over resources binding
- Added a helper for setting viewport and scissor to reuse in in multiple samples

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build and run on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules

Tested on:

- Huawei P20 (Android)
- Desktop (Windows)

### CODE REVIEW

**Comment by JoseEmilio-ARM** [comment]
Please add the sample to the list in https://github.com/KhronosGroup/Vulkan-Samples/blob/master/samples/CMakeLists.txt#L77 (just above AFBC)

**Review by gary-sweet — CHANGES_REQUESTED** [review]
I'm seeing a bunch of repeated validation layer errors with this change:

```
[error] [framework/core/instance.cpp:41] 0 - UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object: 0x3df73d00 (Type = 6) | Submitted command buffer expects VkImage 0x170000000017[] (subresource: aspectMask 0x2 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_UNDEFINED.

[error] [framework/core/instance.cpp:41] 0 - VUID-VkPresentInfoKHR-pImageIndices-01296: [ VUID-VkPresentInfoKHR-pImageIndices-01296 ] Object: 0x3d3a36f0 (Type = 4) | Images passed to present must be in layout VK_IMAGE_LAYOUT_PRESENT_SRC_KHR or VK_IMAGE_LAYOUT_SHARED_PRESENT_KHR but is in VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL. The Vulkan spec states: Each element of pImageIndices must be the index of a presentable image acquired from the swapchain specified by the corresponding element of the pSwapchains array, and the presented image subresource must be in the VK_IMAGE_LAYOUT_PRESENT_SRC_KHR layout at the time the operation is executed on a VkDevice (https://github.com/KhronosGroup/Vulkan-Docs/search?q=VUID-VkPresentInfoKHR-pImageIndices-01296)

[error] [framework/core/instance.cpp:41] 0 - VUID-VkImageMemoryBarrier-oldLayout-01197: [ VUID-VkImageMemoryBarrier-oldLayout-01197 ] Object: 0x3de0dda0 (Type = 6) | For VkImage 0x190000000019[] you cannot transition the layout of aspect=2 level=0 layer=0 from VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_OPTIMAL when the previous known layout is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL. The Vulkan spec states: oldLayout must be VK_IMAGE_LAYOUT_UNDEFINED or the current layout of the image subresources affected by the barrier (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkImageMemoryBarrier-oldLayout-01197)
```

File: samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+Note that the more similar is the workload for all the passes, the more performance improvement you can get by splitting the work between multiple threads. In this sample the same scene is rendered once in each render pass but from different viewpoints and with different complexity of commands recording (shadow pass requires less descriptor and resources setup for each frame). That means the workloads are not fully equal, but it is still worth delegating part of the work to another thread, and the increase in performance is noticeable.
+
+One way to use multi-threading with multiple render passes is to create a separate primary level command buffer for each of them. In this case command buffers can be recorded independently and then submitted to the queue all at once using ``vkQueueSubmit``.
+
+A better approach is to use secondary level command buffers, which helps to reduce the overhead and get even better performance. First, both of the passes are recorded into two separate secondary command buffers using two threads. Then, we can just reference them in the primary command buffer via ``vkCmdExecuteCommands``.
```
Comment (gary-sweet) [inline samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md]: This is an Arm specific statement. For us, the primary buffer mode is slightly faster than secondary.

File: samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+## Best practice summary
+
+**Do**
+
+* Use multi-threading for command buffer recording if possible.
+* Spread the workload between threads as equally, as possible, to utilize them in the most efficient way.
+
+**Avoid**
+
+* Submitting multiple primary command buffers to the queue at once, when secondary command buffers can be used.
```
Comment (gary-sweet) [inline samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md]: Arm specific again

**Review by gary-sweet — CHANGES_REQUESTED** [review]

File: samples/performance/multithreading_render_passes/multithreading_render_passes.cpp [hunk 3]
```diff
@@ -52,7 +52,7 @@ bool MultithreadingRenderPasses::prepare(vkb::Platform &platform)
 		shadow_render_targets[i] = create_shadow_render_target(SHADOWMAP_RESOLUTION);
 	}
 
-	load_scene("scenes/bonza/Bonza4X.gltf");
+	load_scene("scenes/sponza/Sponza01.gltf");
```
Comment (gary-sweet) [inline samples/performance/multithreading_render_passes/multithreading_render_passes.cpp]: Did you intend to change the scene here?
This scene is a lot lighter and I don't really see any performance difference between the modes now.

**Comment by gary-sweet** [comment]
The validation errors appear to be fixed now. Thanks

File: samples/performance/multithreading_render_passes/multithreading_render_passes.cpp (hunk 3 above)
Comment (PavelRudko-ARM) [reply samples/performance/multithreading_render_passes/multithreading_render_passes.cpp]: No, sorry. It's fixed now.

File: samples/performance/multithreading_render_passes/multithreading_render_passes.cpp [hunk 4]
```diff
@@ -349,6 +352,35 @@ void MultithreadingRenderPasses::record_image_memory_barriers(vkb::CommandBuffer
...
+void MultithreadingRenderPasses::record_shadow_pass_image_memory_barrier(vkb::CommandBuffer &command_buffer)
+{
+	auto &shadowmap = shadow_render_targets[render_context->get_active_frame_index()]->get_views().at(shadowmap_attachment_index);
+
+	vkb::ImageMemoryBarrier memory_barrier{};
+	memory_barrier.old_layout      = VK_IMAGE_LAYOUT_UNDEFINED;
+	memory_barrier.new_layout      = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+	memory_barrier.src_access_mask = 0;
+	memory_barrier.dst_access_mask = VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT | VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
+	memory_barrier.src_stage_mask  = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT;
```
Comment (JoseEmilio-ARM) [inline samples/performance/multithreading_render_passes/multithreading_render_passes.cpp:364]: should this also be a fragment stage? @cforfang

Comment (ChristianF-ARM) [reply samples/performance/multithreading_render_passes/multithreading_render_passes.cpp:364]: Looks OK; it’s a WAR hazard (last use of the image was a read some frames ago), so no memory dependency needed, and the execution dependency is satisfied through the fence wait before the images for this frame-index is re-used.

File: samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md (hunk 1 above)
Comment (PavelRudko-ARM) [reply samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md]: I have rephrased it.

File: samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md (hunk 2 above)
Comment (PavelRudko-ARM) [reply samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md]: I have removed it. However, it leaves us with no "Avoid" section. I don't know how bad it is, probably I should think of something else for a substitution.

File: samples/performance/multithreading_render_passes/multithreading_render_passes.cpp [hunk 5]
```diff
@@ -52,7 +52,7 @@ bool MultithreadingRenderPasses::prepare(vkb::Platform &platform)
 		shadow_render_targets[i] = create_shadow_render_target(SHADOWMAP_RESOLUTION);
 	}
 
-	load_scene("scenes/sponza/Sponza01.gltf");
+	load_scene("scenes/bonza/Bonza4x.gltf");
```
Comment (gary-sweet) [inline samples/performance/multithreading_render_passes/multithreading_render_passes.cpp]: This is still wrong I'm afraid. It should be Bonza4X not Bonza4x (upper vs lower case x)

Comment (PavelRudko) [reply samples/performance/multithreading_render_passes/multithreading_render_passes.cpp]: Fixed :)

Comment (PavelRudko-ARM) [reply samples/performance/multithreading_render_passes/multithreading_render_passes.cpp]: Fixed :)

**Comment by r-potter** [comment]
title

It would be good to do something with the title here. Either add some line breaks or shorten it.

File: samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+![Primary Command Buffers](images/primary_command_buffers.png)
+
+With secondary command buffers you can see a further drop in frame time.
+
+![Secondary Command Buffers](images/secondary_command_buffers.png)
+
+## Profiling
+
+Android Profiler can help to see how threads are utilized. Flame Chart shows how much time was spent for each function execution during a particular timeframe. In this particular example total contribution of command buffers recording in the main thread is 9.94 seconds within a 10 seconds capture with multi-threading disabled. 
+
```
Comment (r-potter) [inline samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md:69]: Can we make this a bit more generic and less android specific? i.e. "A profiling tool, such as Android Profiler, can help..."

File: samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+With multi-threading enabled it remains almost the same (9.92s), but in the second thread 5.7s was spent for shadow pass recording. 
+
+![Profiler Capture Secondary Command Buffers](images/android_studio_capture_secondary_buffers.png)
+_Multithreading is enabled and the timeline shows how both threads are used_
+
+That means, two threads perform the same amount of work in 10s as one thread in more than 15.7 seconds and we should expect approximately 1.57 times better performance. 
+
+And indeed, in debug build, which was used for profiling, frame time is decreased from 531.1ms to 337.7ms using multi-threading (1.57 times decrease).
+
```
Comment (r-potter) [inline samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md:81]: If part of the purpose of the samples is to illustrate best practice, we probably shouldn't be using debug builds for profiling.

**Review by r-potter — CHANGES_REQUESTED** [review]

File: samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md (hunk 7 above)
Comment (PavelRudko-ARM) [reply samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md:81]: The purpose of using a debuggable package for profiling was to see function execution times in the Flame Chart. The frame times in the section above, however, were measured in a release build.

**Comment by JoseEmilio-ARM** [comment]
> title
> 
> It would be good to do something with the title here. Either add some line breaks or shorten it.

Shorter term we could remove the 'GPU: ' prefix from the GPU name label to gain a bit more space, and in this case we could perhaps shorten the sample name to 'Multithreading RP'

However this will also be the case for future and present long name samples like 'Specialization constants', so maybe it would be better to raise an issue to fix it in a different PR? Some options like the ones you mentioned are certainly possible but require a bit of fiddling with Imgui to programatically adjust the width and height of the sample name given the width of the screen and the width of the GPU label.

**Review by gary-sweet — DISMISSED** [review]

**Review by r-potter — CHANGES_REQUESTED** [review]
UI/title still overlaps. Shortening the title/GPU name label would help
UI is very sluggish and occasionally doesn't respond due to the long frame durations (350+ms on some devices)
Sample crashes on launch on S7

**Comment by PavelRudko-ARM** [comment]
The title now should not overlap other UI elements.
short_title

**Comment by carlosadcq** [comment]
Tested on Qualcomm platforms. Built properly and ran as expected. I can confirm titles don't overlap anymore.

**Comment by r-potter** [comment]
I was asked to check that I was testing a release build of this due to the sluggish UI performance. 

Release builds appear to trigger an error with the resource caching system:

````
[info] Loaded gltf image #15 (vase_hanging.ktx)
[info] Loaded gltf image #24 (sponza_roof_diff.ktx)
[info] Time spent loading images: 0.043769 seconds across 8 threads.
[error] [ching.h:719] Creation error for #1 cache object (N3vkb12ShaderModuleE)
[error] ['n' exceeds maximum supported size:75] std::exception
[info] ContentRectChanged: 0x7c5a58da80

[info] Closing App (Runtime: 0.7)
[info] Total device memory leaked: 0 bytes.
````

This occurs on both Qualcomm and Mali devices.

**Comment by PavelRudko-ARM** [comment]
> I was asked to check that I was testing a release build of this due to the sluggish UI performance.
> 
> Release builds appear to trigger an error with the resource caching system:
> 
> ```
> [info] Loaded gltf image #15 (vase_hanging.ktx)
> [info] Loaded gltf image #24 (sponza_roof_diff.ktx)
> [info] Time spent loading images: 0.043769 seconds across 8 threads.
> [error] [ching.h:719] Creation error for #1 cache object (N3vkb12ShaderModuleE)
> [error] ['n' exceeds maximum supported size:75] std::exception
> [info] ContentRectChanged: 0x7c5a58da80
> 
> [info] Closing App (Runtime: 0.7)
> [info] Total device memory leaked: 0 bytes.
> ```
> 
> This occurs on both Qualcomm and Mali devices.

I have just rebased to the latest master and tried running it on Huawei P20 with validation layers on. Everything works fine. The error for #1 cache object means an empty fragment shader is not getting compiled. Could you please tell an example of a device? Thanks!

**Comment by r-potter** [comment]
> I have just rebased to the latest master and tried running it on Huawei P20 with validation layers on. Everything works fine. The error for #1 cache object means an empty fragment shader is not getting compiled. Could you please tell an example of a device? Thanks!

I saw it on both an Adreno 630 (Note 9) and a Mali-based S10

**Comment by PavelRudko-ARM** [comment]
> > I have just rebased to the latest master and tried running it on Huawei P20 with validation layers on. Everything works fine. The error for #1 cache object means an empty fragment shader is not getting compiled. Could you please tell an example of a device? Thanks!
> 
> I saw it on both an Adreno 630 (Note 9) and a Mali-based S10

I have just tested it on an S10 with Mali and the sample works fine. Could you, please, make sure you test it with up-to-date sources and assets? I have rebased to latest master, so it should be synced now.

**Comment by r-potter** [comment]
A completely clean checkout appears to work fine. The sluggish performance is indeed only on debug builds. Release builds give a more reasonable 25-40ms per frame. I'm happy with this MR now.

**Review by r-potter — APPROVED** [review]

**Review by gary-sweet — APPROVED** [review]

**Review by AlejandroCosin — APPROVED** [review]

**Review by carlosadcq — APPROVED** [review]

**Pull request merged**